fdbserver/ConfigBroadcaster.actor.cpp

```diff
@@ -192,34 +199,48 @@ class ConfigBroadcasterImpl {
 		return Void();
 	}
 
-	ACTOR static Future<Void> waitForFailure(ConfigBroadcasterImpl* self,
+	ACTOR static Future<Void> waitForFailure(ConfigBroadcasterImpl* self, Future<Void> watcher, UID clientUID) {
+		wait(watcher);
+		TraceEvent(SevDebug, "ConfigBroadcastClientDied", self->id).detail("ClientID", clientUID);
+		self->clients.erase(clientUID);
+		return Void();
+	}
+
+	ACTOR static Future<Void> registerWorker(ConfigBroadcaster* self,
+	                                         ConfigBroadcasterImpl* impl,
+	                                         Version lastSeenVersion,
+	                                         ConfigClassSet configClassSet,
 	                                         Future<Void> watcher,
-	                                         BroadcastClientDetails* client) {
-		wait(success(watcher));
-		self->clients.erase(std::remove(self->clients.begin(), self->clients.end(), *client));
+	                                         ConfigBroadcastInterface broadcastInterface) {
+		state BroadcastClientDetails client(
+		    watcher, std::move(configClassSet), lastSeenVersion, std::move(broadcastInterface));
+		if (!impl->consumerFuture.isValid()) {
+			impl->consumerFuture = impl->consumer->consume(*self);
+		}
+
+		if (impl->clients.count(broadcastInterface.id())) {
+			// Client already registered
+			return Void();
+		}
+
+		TraceEvent(SevDebug, "ConfigBroadcasterRegisteringWorker", impl->id)
+		    .detail("ClientID", broadcastInterface.id())
+		    .detail("MostRecentVersion", impl->mostRecentVersion)
+		    .detail("ClientLastSeenVersion", lastSeenVersion);
+		// Push full snapshot to worker if it isn't up to date.
+		wait(impl->pushSnapshot(impl->mostRecentVersion, client));
+		impl->clients[broadcastInterface.id()] = client;
+		impl->actors.add(waitForFailure(impl, watcher, broadcastInterface.id()));
 		return Void();
 	}
 
 public:
-	Future<Void> registerWorker(ConfigBroadcaster* self,
+	Future<Void> registerWorker(ConfigBroadcaster& self,
```

Comment:
One question -- why change this parameter to be a reference when it requires dereferencing `this` at the call site and then passing an address to the `registerWorker` actor below?

Reply:
The reference should always be valid when `registerWorker` is called, so prevents calling something like `registerWorker(nullptr, ...)`. Unfortunately the actor compiler requires passing a pointer to the static ACTOR function.